[programmable transactions] Add scaffold for `Upgrade` command

tzakian opened the pull request:
This adds the basic plumbing for the `Upgrade` command. It takes an (owned) `UpgradeTicket` as an argument and will produce a `ReceiptTicket` with the new package's ID along with the `UpgradeTicket`'s `UpgradeCap` ID.  

There's probably some places where I did stupid stuff since I'm still paging things in, so please let me know if I went off-track somewhere :)

tzakian commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -404,6 +407,39 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>>(
+    let modules = publish_and_verify_modules(context, module_bytes)?;
+    let package_id = context.new_package(modules)?;
+
+    // TODO(tzakian): check digest, that the package ID points to a package, and link compatibility
```
It looks like the `ExecutionContext` has a `StateView` which will allow us to read form the object store for this purpose. I believe this should give us what we want :)

tnowacki replied:
The state view is the `TemporaryStore`. I think you would need to somehow expose the object store directly? If I'm understanding correctly

amnn replied:
That's right, `TemporaryStore` is only going to include the input objects for the transaction + any objects that we dynamically load (could we use the `ObjectRuntime` to do this @tnowacki?) so you won't find the package to be upgraded in there by default.  In most cases, packages are not loaded this way, they are loaded when `MoveVM` asks to load a module (which goes via `SyncModuleCache`).

tnowacki commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -404,6 +407,39 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>>(
+    upgrade_ticket_arg: Argument,
+    module_bytes: Vec<Vec<u8>>,
+) -> Result<Value, ExecutionError> {
+    assert_invariant!(
```
The config is in the `Context`, so it should be easy to check. We should add/have some error in `ExecutionErrorKind` for `NotYetSupported` or something similar

tnowacki commented on crates/sui-adapter/src/programmable_transactions/context.rs:
```diff
@@ -365,6 +365,11 @@ where
         Ok(id)
     }
 
+    /// Add a package to the context.
+    pub fn add_package(&mut self, package_object: Object) {
+        self.new_packages.push(package_object)
+    }
+
```
If you are going to do this, could you refactor `new_package` out? I'd rather not have two functions that push to `new_packages`.

tnowacki commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -271,6 +274,14 @@ fn execute_command<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
             execute_move_publish::<_, _, Mode>(context, &mut argument_updates, modules)?
         }
+        Command::Upgrade(upgrade_ticket, dep_ids, modules) => {
+            vec![execute_move_upgrade(
```
nit, make `execute_move_upgrade` return a `Vec`, just to be a bit more inline/idiomatic

tnowacki commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -436,6 +447,92 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
+        .protocol_config
+        .check_package_upgrades_supported()
+        .map_err(|_| {
+            ExecutionError::from_kind(ExecutionErrorKind::PackageUpgradeError(
```
I think we probably just want a general error for "this feature isn't supported". Might be a bad idea to start carving it out per feature.

tnowacki commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -436,6 +447,92 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
+        let mut ticket_bytes = Vec::new();
+        let ticket_obj_val: ObjectValue =
+            context.by_value_arg(CommandKind::Upgrade, 0, upgrade_ticket_arg)?;
+        ticket_obj_val.write_bcs_bytes(&mut ticket_bytes);
```
You need to type check the `ticket_obj_val` before doing `bcs::from_bytes` or you are going to accidentally support type conversion for upgrade tickets... which would be bad :)

tnowacki commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -436,6 +447,92 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
+    context: &'a ExecutionContext<E, S>,
+    package_id: &ObjectID,
+) -> Result<&'a MovePackage, ExecutionError> {
+    let object = context.state_view.read_object(package_id).ok_or_else(|| {
```
This will read from the temporary store, but no other new packages should be in there yet. Is that okay?

tzakian replied:
Not sure, so I'll let @amnn weigh-in. 

Can you publish multiple packages in the same transaction? And even if so I'm guessing this would be OK since they should have different package IDs I'd suspect...

amnn replied:
You can indeed publish multiple packages in the same transaction, and that could cause an issue here, where someone publishes and then upgrades a package in one transaction.  It's a strange thing to do, but we should respond to it in a sensible way.  In this case, it should be possible to support by also checking the `context`'s list of `new_packages`, right (which may then need to become an `IndexMap`.

But @tnowacki, will the temporary store be able to read the original package being upgraded in the first place? It's not going to be a transaction input.

tnowacki commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -481,6 +578,60 @@ fn vm_move_call<E: fmt::Debug, S: StorageView<E>>(
     Ok(result)
 }
 
+fn upgrade_check_compatiblity_and_publish_package<E: fmt::Debug, S: StorageView<E>>(
```
Is this more of a "signature" check? Not sure where this should live or not

tnowacki commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -481,6 +578,60 @@ fn vm_move_call<E: fmt::Debug, S: StorageView<E>>(
+    let upgraded_pkg = upgraded_package_object
+        .data
+        .try_as_package_mut()
+        .ok_or_else(|| ExecutionError::from_kind(ExecutionErrorKind::InvariantViolation))?;
```
nit `invariant_violation!("A helpful message for your future self oncall")`

tnowacki commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -481,6 +578,60 @@ fn vm_move_call<E: fmt::Debug, S: StorageView<E>>(
+    // The current package should be able to be normalized otherwise something has gone terribly wrong.
+    let current_normalized = current_package
+        .normalize()
+        .map_err(|_| ExecutionError::from_kind(ExecutionErrorKind::InvariantViolation))?;
```
ditto

tnowacki commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -481,6 +578,60 @@ fn vm_move_call<E: fmt::Debug, S: StorageView<E>>(
+    let current_normalized = current_package
+        .normalize()
+        .map_err(|_| ExecutionError::from_kind(ExecutionErrorKind::InvariantViolation))?;
+    let mut new_normalized = upgraded_pkg.normalize().map_err(|_| {
```
When does normalize error? this doesn't feel like the right error kind below

tzakian replied:
Good spot! This only happens if the deserialization of the modules fail. I'll rework this and it looks like we can get rid of the `Result` all together.

tnowacki commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -481,6 +578,60 @@ fn vm_move_call<E: fmt::Debug, S: StorageView<E>>(
+            return Err(ExecutionError::from_kind(ExecutionErrorKind::PackageUpgradeError(PackageUpgradeError::IncompatibleUpgrade)))
+        };
+
+        if compatibility.check(&cur_module, &new_module).is_err() {
```
What is the error here? Does it give more info? Should we pass that along to the user?
Maybe at least we could give it as a string for dev-inspect cases

tnowacki replied:
`ExecutionError::new_with_source` is the function for doing so

tnowacki commented on crates/sui-types/src/move_package.rs:
```diff
@@ -88,6 +90,22 @@ pub struct UpgradeCap {
+/// Rust representation of `sui::package::UpgradeTicket`.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct UpgradeTicket {
+    pub cap: ID,
```
what is this field?

tnowacki replied:
Maybe I'm just a fool and should look at the `sui::package` code lol

amnn replied:
https://github.com/MystenLabs/sui/blob/73dcf13605dcbec42646b75451245794946da155/crates/sui-framework/sources/package.move#L57-L76

amnn commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -436,6 +446,93 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
+    };
+
+    // Run the move + sui verifier on the modules and publish them into the cache.
+    let upgraded_package_modules = publish_and_verify_modules::<_, _, Mode>(context, module_bytes)?;
```
We shouldn't do this until we have checked compatibility, otherwise we're going to pollute the loader unnecessarily.

tzakian replied:
Sounds good. I originally had it lower down, but lifetime issues made me want to move it up here. I'll fix it :)

tzakian replied:
Actually one question I had here is do we want to run the normalizer before we perform verification? From what I could tell poking around the code it seemed like a "this is fine, but you should really run the bytecode verifier first before doing this" which would mean we'd want to verify and publish first?

amnn commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -481,6 +578,54 @@ fn vm_move_call<E: fmt::Debug, S: StorageView<E>>(
+
+    // The current package should be able to be deserialized and normalized otherwise something has gone terribly wrong.
+    let Ok(current_normalized) = current_package.normalize() else {
+            invariant_violation!("Tried to normalize modules in already-published package but failed")
```
nit, weird indent.

tzakian replied:
Didn't realize cargo fmt doesn't format `let-else` yet!

amnn commented on crates/sui-adapter/src/programmable_transactions/execution.rs:
```diff
@@ -481,6 +578,54 @@ fn vm_move_call<E: fmt::Debug, S: StorageView<E>>(
+
+    for (name, cur_module) in current_normalized {
+        let Some(new_module) = new_normalized.remove(&name) else {
+            return Err(ExecutionError::from_kind(ExecutionErrorKind::PackageUpgradeError(PackageUpgradeError::IncompatibleUpgrade)))
```
nit: Is there a way for us to explain the nature of this error, like we do below for the individual compatibility issues?

tzakian replied:
I'm pretty sure we can. `ExistingModuleNotInUpgradedPackage(Identifier)` or something like that with a corresponding explanation of "Existing module {name} in package not found in next version of package" (or something like that).

amnn commented on crates/sui-protocol-config/src/lib.rs:
```diff
@@ -113,6 +113,7 @@ pub struct Error(pub String);
 struct FeatureFlags {
     // Add feature flags here, e.g.:
     // new_protocol_feature: bool,
+    package_upgrades_allowed: bool,
```
tiny nit: Do we need the `_allowed`? `package_upgrades` is the feature.
```suggestion
    package_upgrades: bool,
```

tzakian replied:
I don't think we do. I'll rename it :)

amnn reviewed (commented):
Looking great!  I would say, initially, prioritise getting the change to the `Command` interface out, along with the protocol config change and a test to show that if you try and call the upgrade command, it will fail.  We can follow up with the rest of the implementation and testing in another PR which won't need to go in before March 15th.

tzakian commented:
I'm going to split this PR into two: the scaffolding part (which we do need to get in ASAP) and the actual implementation. Once I do that I'll undraft this PR.

tzakian commented:
Split PR so this contains only the skeleton/scaffolding. The next PR will be working on the actual upgrade logic. 

Added a simple Rust test to make sure we fail with the correct error if we try and upgrade a package right now.

amnn commented on crates/sui-types/src/move_package.rs:
```diff
@@ -32,6 +32,8 @@ use std::collections::BTreeMap;
 
 pub const PACKAGE_MODULE_NAME: &IdentStr = ident_str!("package");
 pub const UPGRADECAP_STRUCT_NAME: &IdentStr = ident_str!("UpgradeCap");
+pub const UPGRADETICKET_STRUCT_NAME: &IdentStr = ident_str!("UpgradeTicket");
```
I think the changes to this file could technically go into your second PR, but not a big deal.

amnn reviewed (approved):
🚀  Yes!  This is great, I think this is good to go, awesome work @tzakian, thank you for the quick turnaround cc @rvantonder to crib off this PR for making the relevant changes to the `Publish` command. @awelc will take it from there to interpret the `deps` coming in during publish and conver them to a linkage table!

The aim is that by the upcoming cut, a package that is published has all the right things (linkage and type origin table) in place so that it can be upgraded once we land the support for that in future.
